src/settings/README.md

```diff
@@ -0,0 +1,96 @@
+
+We don't use reflections because it is very troublesome in C++ and it is not compilation time check. A wrong parameter name will not be discovered until run time.
+
+Writing the `enum` is made much easier thanks to the macro based design of the Settings Manager, which is explained below. 
+
+### Macro based Parameter Definition
+
+We provide some macros for the developers to define their parameters. They include `SETTING_int`, `SETTING_bool`, `SETTING_double` and `SETTING_string`. The definition of the macros is in `settings_macro.h`.
+
+Although macros are confusing and not straightforward, it offers more benefits than disadvantages in the Settings Manager. The biggest benefit is that the users can now specify each parameter only once. Without macros, when a user wants to define a new parameter, he must do it for three times: append it in the `Params` `enum`, register it in `gflags`, and initialize it in the Settings Manager. It is clear that this exposes too many internal details of the Settings Manager to the users. 
+
+In the current design, users only need to define his parameter in `settings.h` only once, and the Settings Manager will do the `enum`, `gflags` and initialization stuff for them by simply `#include settings.h` and interpret `SETTING_xxx` in different ways.
+
+### Internal and External Map
+
+Although there is a physical `pg_settings` table in the catalogs where we can find all the information about a parameter, we finally decided that there should be a map that contains the values of parameters. Why? Because:
+
+- The most important reason is that accessing the catalog needs to be done in a transaction. This can cause infinite loops. A typical situation is, on system bootstrapping, the Settings Manager needs to open a transaction to append the values into the `pg_settings` table. However, as long as there is a configuration needed by the Transaction Manager (a very likely one can be the default isolation level), it will call `get` to ask for that parameter, and the Settings Manager will open a new transaction to read the parameter from the catalog, which results in an infinite loop. In order to reduce the interaction with other parts of the system, we decided to maintain a map inside the Settings Manager. In this way, it can serve all requests without consulting anyone else.
+- It is much easier to operate on a map than the catalog. 
+- Since the number of parameters is low, an extra map won't introduce much overhead.
+
+- Although this essentially turned the Settings Manager as a cache layer for the `pg_settings` table, its consistency is easy to maintain (just update the catalog in `set` methods).
+
+This value map is an external map actually stored in DBMain. We also have another internal callback map, which maps a setting to its callback. It should be immutable during runtime.
+
+### Synchronous Modification Enforcement
```

Comment:
How does async call maintain consistency?

Reply:
Modification in an asynchronous manner with callback itself will not have consistency issue. Because the observer of any setting value are not changed during the whole process. Namely, consistency issue happens when there are multiple observers (or owners) of a value, which is not what we are explaining here.

If you are asking how to maintain consistency between different observers, which would happen no matter the modification is async or not, you can refer to the sections above :)